graphql/src/store/prefetch.rs

```diff
@@ -0,0 +1,752 @@
+            // an empty `Vec`)
+            // This is complicated by the fact that, if there was a type
+            // condition, we should only change parents that meet the type
+            // condition; we set it for all parents regardless, as that does
+            // not cause problems in later processing, but make sure that we
+            // do not clobber an entry under this `response_key` that might
+            // have been set by a previous join by appending values rather
+            // than using straight insert into the parent
+            let mut values = parent
+                .get(self.parent_field)
+                .and_then(|key| match key {
+                    StoreValue::String(key) => {
+                        grouped.get(key.as_str()).map(|values| values.clone())
+                    }
+                    StoreValue::List(keys) => {
+                        // Note that each `grouped.get` returns a vec with
+                        // at most one element, and since `key` is the id
+                        // of that element, the resulting `values` will not
+                        // contain duplicates
+                        Some(
+                            keys.iter()
+                                .filter_map(|key| key.as_str())
+                                .filter_map(|key| grouped.get(key))
+                                .flatten()
+                                .map(|rc| rc.clone())
+                                .collect::<Vec<_>>(),
+                        )
+                    }
+                    StoreValue::Null => None,
+                    _ => unreachable!("join fields must be strings or lists of strings"),
+                })
+                .unwrap_or(vec![]);
+            parent
+                .children
+                .entry(response_key.to_owned())
+                .or_default()
+                .append(&mut values);
+        }
+    }
+}
+
+/// Run the query in `ctx` in such a manner that we only perform one query
+/// per 'level' in the query. A query like `musicians { id bands { id } }`
+/// will perform two queries: one for musicians, and one for bands, regardless
+/// of how many musicians there are.
+///
+/// The returned value contains a `q::Value::Object` that contains a tree of
+/// all the entities (converted into objects) in the form in which they need
+/// to be returned. Nested object fields appear under the key `r:response_key`
+/// in these objects, and are always `q::Value::List` of objects. In addition,
+/// each entity has a property `:prefetch` set so that the resolver can
+/// tell whether the `r:response_key` associations should be there.
+///
+/// For the above example, the returned object would have one entry under
+/// `r:musicians`, which is a list of all the musicians; each musician has an
+/// entry `r:bands` that contains a list of the bands for that musician. Note
+/// that even for single-object fields, we return a list
```

Comment:
It's because we try to select 2 entries so we can spot if the data in the store has an invalid relationship (same as what the existing resolver does in `resolve_object`)

Expanded the comment